File: fraud_risk_scoring.py
"""
Fraud Risk Scoring System for Financial Transactions.

Analyzes each transaction independently, identifies risk signals,
computes a risk score (0-100), assigns a risk category (LOW, MEDIUM, HIGH),
and generates a transaction-level risk report.
"""

import pandas as pd


def load_dataset(filepath: str) -> pd.DataFrame:
    """Load the transaction dataset from a CSV file."""
    df = pd.read_csv(filepath)
    df.index.name = "transaction_id"
    df = df.reset_index()
    return df


def compute_amount_risk(amount: float) -> tuple[float, str]:
    """
    Compute risk contribution from transaction amount.

    Transactions above 10,000 are considered high risk.
    Score contribution scales with amount magnitude.
    """
    if amount > 500000:
        return 25.0, "Very high transaction amount (>{:,.2f})".format(amount)
    elif amount > 200000:
        return 20.0, "Extremely large transaction amount ({:,.2f})".format(amount)
    elif amount > 100000:
        return 15.0, "Very large transaction amount ({:,.2f})".format(amount)
    elif amount > 10000:
        return 10.0, "High transaction amount ({:,.2f})".format(amount)
    elif amount > 5000:
        return 5.0, "Moderate transaction amount ({:,.2f})".format(amount)
    return 0.0, ""


def compute_type_risk(txn_type: str) -> tuple[float, str]:
    """
    Compute risk contribution from transaction type.

    CASH_OUT and TRANSFER are higher risk transaction types.
    """
    risk_map = {
        "CASH_OUT": (20.0, "High-risk transaction type: CASH_OUT"),
        "TRANSFER": (15.0, "High-risk transaction type: TRANSFER"),
        "DEBIT": (5.0, "Moderate-risk transaction type: DEBIT"),
        "PAYMENT": (0.0, ""),
        "CASH_IN": (0.0, ""),
    }
    return risk_map.get(txn_type, (0.0, ""))


def compute_balance_anomaly_risk(row: pd.Series) -> tuple[float, str]:
    """
    Detect unusual balance patterns that may indicate fraud.

    Checks for:
    - Account being drained (new balance is 0 after transaction)
    - Balance mismatch (old balance - amount != new balance)
    - Destination balance not changing despite receiving funds
    """
    reasons = []
    score = 0.0

    old_bal = row["oldbalanceOrg"]
    new_bal = row["newbalanceOrig"]
    amount = row["amount"]

    # Account drained to zero
    if old_bal > 0 and new_bal == 0:
        score += 10.0
        reasons.append("Origin account fully drained to zero balance")

    # Balance discrepancy at origin
    expected_new_bal = old_bal - amount
    if abs(expected_new_bal - new_bal) > 0.01 and old_bal > 0:
        score += 5.0
        reasons.append(
            "Balance discrepancy at origin (expected {:.2f}, got {:.2f})".format(
                expected_new_bal, new_bal
            )
        )

    # Origin account has zero balance before transaction
    if old_bal == 0 and amount > 0:
        score += 5.0
        reasons.append("Transaction from account with zero initial balance")

    # Destination balance anomaly: balance doesn't increase after receiving
    old_dest = row["oldbalanceDest"]
    new_dest = row["newbalanceDest"]
    name_dest = str(row["nameDest"])
    if not name_dest.startswith("M"):  # Exclude merchants
        if old_dest > 0 and new_dest == 0:
            score += 5.0
            reasons.append("Destination balance dropped to zero after receiving funds")

    return min(score, 20.0), "; ".join(reasons)


def compute_repeat_account_risk(
    df: pd.DataFrame,
) -> dict[int, tuple[float, str]]:
    """
    Identify repeated transactions from the same originating account.

    Rapid sequences of transactions from the same account increase risk.
    """
    account_counts = df.groupby("nameOrig").size()
    repeat_accounts = account_counts[account_counts > 1].index.tolist()

    risk_map = {}
    for idx, row in df.iterrows():
        if row["nameOrig"] in repeat_accounts:
            count = account_counts[row["nameOrig"]]
            score = min(count * 5.0, 15.0)
            risk_map[idx] = (
                score,
                "Account {} has {} transactions in dataset (repeated activity)".format(
                    row["nameOrig"], count
                ),
            )
        else:
            risk_map[idx] = (0.0, "")

    return risk_map


def compute_destination_risk(df: pd.DataFrame) -> dict[int, tuple[float, str]]:
    """
    Assess risk based on destination account patterns.

    Transactions to frequently targeted destination accounts are riskier.
    """
    dest_counts = df.groupby("nameDest").size()
    high_traffic_dests = dest_counts[dest_counts > 2].index.tolist()

    risk_map = {}
    for idx, row in df.iterrows():
        name_dest = str(row["nameDest"])
        if name_dest.startswith("M"):
            # Merchants are generally less risky
            risk_map[idx] = (0.0, "")
        elif name_dest in high_traffic_dests:
            count = dest_counts[name_dest]
            score = min(count * 3.0, 10.0)
            risk_map[idx] = (
                score,
                "Destination {} received {} transactions (high-traffic destination)".format(
                    name_dest, count
                ),
            )
        else:
            risk_map[idx] = (0.0, "")

    return risk_map


def compute_cashout_pattern_risk(df: pd.DataFrame) -> dict[int, tuple[float, str]]:
    """
    Detect high amount followed by cash-out patterns.

    Fraudulent transactions often involve large transfers followed by
    immediate cash-out from the destination account.
    """
    risk_map = {idx: (0.0, "") for idx in df.index}

    # Group by step to find transactions in the same time period
    cashout_dests = set(df[df["type"] == "CASH_OUT"]["nameOrig"].tolist())
    transfer_rows = df[
        (df["type"].isin(["TRANSFER", "CASH_OUT"])) & (df["amount"] > 10000)
    ]

    for idx, row in transfer_rows.iterrows():
        if row["type"] == "TRANSFER" and row["nameDest"] in cashout_dests:
            risk_map[idx] = (
                10.0,
                "Large transfer to account that also performs cash-out (potential layering)",
            )
        elif row["type"] == "CASH_OUT" and row["amount"] > 50000:
            risk_map[idx] = (
                10.0,
                "Large cash-out transaction ({:,.2f}), potential fraud cash-out".format(
                    row["amount"]
                ),
            )

    return risk_map


def assign_risk_level(score: float) -> str:
    """Assign risk category based on score thresholds."""
    if score > 70:
        return "HIGH"
    elif score >= 40:
        return "MEDIUM"
    return "LOW"


def generate_risk_report(df: pd.DataFrame) -> pd.DataFrame:
    """
    Generate a comprehensive transaction-level risk report.

    Combines all risk signals to compute a final risk score and category
    for each transaction.
    """
    # Pre-compute aggregate risk signals
    repeat_risk = compute_repeat_account_risk(df)
    dest_risk = compute_destination_risk(df)
    cashout_risk = compute_cashout_pattern_risk(df)

    results = []

    for idx, row in df.iterrows():
        explanations = []
        total_score = 0.0

        # 1. Amount risk
        amt_score, amt_reason = compute_amount_risk(row["amount"])
        total_score += amt_score
        if amt_reason:
            explanations.append(amt_reason)

        # 2. Transaction type risk
        type_score, type_reason = compute_type_risk(row["type"])
        total_score += type_score
        if type_reason:
            explanations.append(type_reason)

        # 3. Balance anomaly risk
        bal_score, bal_reason = compute_balance_anomaly_risk(row)
        total_score += bal_score
        if bal_reason:
            explanations.append(bal_reason)

        # 4. Repeated account risk
        rep_score, rep_reason = repeat_risk[idx]
        total_score += rep_score
        if rep_reason:
            explanations.append(rep_reason)

        # 5. Destination risk
        dst_score, dst_reason = dest_risk[idx]
        total_score += dst_score
        if dst_reason:
            explanations.append(dst_reason)

        # 6. Cash-out pattern risk
        co_score, co_reason = cashout_risk[idx]
        total_score += co_score
        if co_reason:
            explanations.append(co_reason)

        # 7. Known fraud indicator boost
        if row.get("isFraud", 0) == 1:
            total_score += 15.0
            explanations.append("Transaction flagged as fraud in dataset")

        if row.get("isFlaggedFraud", 0) == 1:
            total_score += 10.0
            explanations.append("Transaction flagged by business fraud detection rules")

        # Cap score at 100
        final_score = min(round(total_score, 2), 100.0)
        risk_level = assign_risk_level(final_score)
        explanation = (
            "; ".join(explanations)
            if explanations
            else "No significant risk signals detected"
        )

        results.append(
            {
                "transaction_id": row["transaction_id"],
                "risk_score": final_score,
                "risk_level": risk_level,
                "explanation": explanation,
            }
        )

    return pd.DataFrame(results)


def main() -> None:
    """Main entry point for the fraud risk scoring system."""
    input_path = "data/Example1.csv"
    output_path = "data/fraud_risk_report.csv"

    print("Loading dataset from '{}'...".format(input_path))
    df = load_dataset(input_path)
    print("Loaded {} transactions.".format(len(df)))

    print("\nAnalyzing transactions for fraud risk signals...")
    report = generate_risk_report(df)

    # Summary statistics
    print("\n--- Risk Score Summary ---")
    print("Total transactions analyzed: {}".format(len(report)))
    print(
        "Risk level distribution:\n{}".format(
            report["risk_level"].value_counts().to_string()
        )
    )
    print(
        "\nRisk score statistics:\n{}".format(
            report["risk_score"].describe().to_string()
        )
    )

    # Show HIGH risk transactions
    high_risk = report[report["risk_level"] == "HIGH"]
    if not high_risk.empty:
        print("\n--- HIGH Risk Transactions ({}) ---".format(len(high_risk)))
        for _, row in high_risk.iterrows():
            print(
                "  Transaction {}: score={}, explanation={}".format(
                    row["transaction_id"], row["risk_score"], row["explanation"]
                )
            )

    # Save report
    report.to_csv(output_path, index=False)
    print("\nRisk report saved to '{}'".format(output_path))


if __name__ == "__main__":
    main()

File: java/src/main/java/com/fraudrisk/FraudRiskScoring.java
package com.fraudrisk;

import com.opencsv.CSVReader;
import com.opencsv.CSVWriter;
import com.opencsv.exceptions.CsvValidationException;

import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.StringJoiner;

/**
 * Fraud Risk Scoring System for Financial Transactions.
 *
 * Analyzes each transaction independently, identifies risk signals,
 * computes a risk score (0-100), assigns a risk category (LOW, MEDIUM, HIGH),
 * and generates a transaction-level risk report.
 */
public class FraudRiskScoring {

    /**
     * Load the transaction dataset from a CSV file.
     */
    public List<Transaction> loadDataset(String filepath) throws IOException, CsvValidationException {
        List<Transaction> transactions = new ArrayList<>();
        try (CSVReader reader = new CSVReader(new FileReader(filepath))) {
            String[] header = reader.readNext(); // skip header
            if (header == null) {
                return transactions;
            }

            int transactionId = 0;
            String[] line;
            while ((line = reader.readNext()) != null) {
                Transaction txn = new Transaction();
                txn.setTransactionId(transactionId);
                txn.setStep(Integer.parseInt(line[0].trim()));
                txn.setType(line[1].trim());
                txn.setAmount(Double.parseDouble(line[2].trim()));
                txn.setNameOrig(line[3].trim());
                txn.setOldbalanceOrg(Double.parseDouble(line[4].trim()));
                txn.setNewbalanceOrig(Double.parseDouble(line[5].trim()));
                txn.setNameDest(line[6].trim());
                txn.setOldbalanceDest(Double.parseDouble(line[7].trim()));
                txn.setNewbalanceDest(Double.parseDouble(line[8].trim()));
                txn.setIsFraud(Integer.parseInt(line[9].trim()));
                txn.setIsFlaggedFraud(Integer.parseInt(line[10].trim()));
                transactions.add(txn);
                transactionId++;
            }
        }
        return transactions;
    }

    /**
     * Compute risk contribution from transaction amount.
     *
     * Transactions above 10,000 are considered high risk.
     * Score contribution scales with amount magnitude.
     */
    public RiskResult computeAmountRisk(double amount) {
        if (amount > 500000) {
            return new RiskResult(25.0, String.format("Very high transaction amount (>%,.2f)", amount));
        } else if (amount > 200000) {
            return new RiskResult(20.0, String.format("Extremely large transaction amount (%,.2f)", amount));
        } else if (amount > 100000) {
            return new RiskResult(15.0, String.format("Very large transaction amount (%,.2f)", amount));
        } else if (amount > 10000) {
            return new RiskResult(10.0, String.format("High transaction amount (%,.2f)", amount));
        } else if (amount > 5000) {
            return new RiskResult(5.0, String.format("Moderate transaction amount (%,.2f)", amount));
        }
        return new RiskResult(0.0, "");
    }

    /**
     * Compute risk contribution from transaction type.
     *
     * CASH_OUT and TRANSFER are higher risk transaction types.
     */
    public RiskResult computeTypeRisk(String txnType) {
        switch (txnType) {
            case "CASH_OUT":
                return new RiskResult(20.0, "High-risk transaction type: CASH_OUT");
            case "TRANSFER":
                return new RiskResult(15.0, "High-risk transaction type: TRANSFER");
            case "DEBIT":
                return new RiskResult(5.0, "Moderate-risk transaction type: DEBIT");
            case "PAYMENT":
            case "CASH_IN":
                return new RiskResult(0.0, "");
            default:
                return new RiskResult(0.0, "");
        }
    }

    /**
     * Detect unusual balance patterns that may indicate fraud.
     *
     * Checks for:
     * - Account being drained (new balance is 0 after transaction)
     * - Balance mismatch (old balance - amount != new balance)
     * - Destination balance not changing despite receiving funds
     */
    public RiskResult computeBalanceAnomalyRisk(Transaction txn) {
        List<String> reasons = new ArrayList<>();
        double score = 0.0;

        double oldBal = txn.getOldbalanceOrg();
        double newBal = txn.getNewbalanceOrig();
        double amount = txn.getAmount();

        // Account drained to zero
        if (oldBal > 0 && newBal == 0) {
            score += 10.0;
            reasons.add("Origin account fully drained to zero balance");
        }

        // Balance discrepancy at origin
        double expectedNewBal = oldBal - amount;
        if (Math.abs(expectedNewBal - newBal) > 0.01 && oldBal > 0) {
            score += 5.0;
            reasons.add(String.format("Balance discrepancy at origin (expected %.2f, got %.2f)",
                    expectedNewBal, newBal));
        }

        // Origin account has zero balance before transaction
        if (oldBal == 0 && amount > 0) {
            score += 5.0;
            reasons.add("Transaction from account with zero initial balance");
        }

        // Destination balance anomaly: balance doesn't increase after receiving
        double oldDest = txn.getOldbalanceDest();
        double newDest = txn.getNewbalanceDest();
        String nameDest = txn.getNameDest();
        if (!nameDest.startsWith("M")) { // Exclude merchants
            if (oldDest > 0 && newDest == 0) {
                score += 5.0;
                reasons.add("Destination balance dropped to zero after receiving funds");
            }
        }

        return new RiskResult(Math.min(score, 20.0), String.join("; ", reasons));
    }

    /**
     * Identify repeated transactions from the same originating account.
     *
     * Rapid sequences of transactions from the same account increase risk.
     */
    public Map<Integer, RiskResult> computeRepeatAccountRisk(List<Transaction> transactions) {
        // Count transactions per originating account
        Map<String, Integer> accountCounts = new HashMap<>();
        for (Transaction txn : transactions) {
            accountCounts.merge(txn.getNameOrig(), 1, Integer::sum);
        }

        Map<Integer, RiskResult> riskMap = new HashMap<>();
        for (int i = 0; i < transactions.size(); i++) {
            Transaction txn = transactions.get(i);
            int count = accountCounts.getOrDefault(txn.getNameOrig(), 0);
            if (count > 1) {
                double score = Math.min(count * 5.0, 15.0);
                riskMap.put(i, new RiskResult(score,
                        String.format("Account %s has %d transactions in dataset (repeated activity)",
                                txn.getNameOrig(), count)));
            } else {
                riskMap.put(i, new RiskResult(0.0, ""));
            }
        }
        return riskMap;
    }

    /**
     * Assess risk based on destination account patterns.
     *
     * Transactions to frequently targeted destination accounts are riskier.
     */
    public Map<Integer, RiskResult> computeDestinationRisk(List<Transaction> transactions) {
        // Count transactions per destination account
        Map<String, Integer> destCounts = new HashMap<>();
        for (Transaction txn : transactions) {
            destCounts.merge(txn.getNameDest(), 1, Integer::sum);
        }

        Map<Integer, RiskResult> riskMap = new HashMap<>();
        for (int i = 0; i < transactions.size(); i++) {
            Transaction txn = transactions.get(i);
            String nameDest = txn.getNameDest();
            if (nameDest.startsWith("M")) {
                // Merchants are generally less risky
                riskMap.put(i, new RiskResult(0.0, ""));
            } else {
                int count = destCounts.getOrDefault(nameDest, 0);
                if (count > 2) {
                    double score = Math.min(count * 3.0, 10.0);
                    riskMap.put(i, new RiskResult(score,
                            String.format("Destination %s received %d transactions (high-traffic destination)",
                                    nameDest, count)));
                } else {
                    riskMap.put(i, new RiskResult(0.0, ""));
                }
            }
        }
        return riskMap;
    }

    /**
     * Detect high amount followed by cash-out patterns.
     *
     * Fraudulent transactions often involve large transfers followed by
     * immediate cash-out from the destination account.
     */
    public Map<Integer, RiskResult> computeCashoutPatternRisk(List<Transaction> transactions) {
        Map<Integer, RiskResult> riskMap = new HashMap<>();
        for (int i = 0; i < transactions.size(); i++) {
            riskMap.put(i, new RiskResult(0.0, ""));
        }

        // Collect accounts that perform CASH_OUT
        Set<String> cashoutDests = new HashSet<>();
        for (Transaction txn : transactions) {
            if ("CASH_OUT".equals(txn.getType())) {
                cashoutDests.add(txn.getNameOrig());
            }
        }

        // Check transfer rows with amount > 10000
        for (int i = 0; i < transactions.size(); i++) {
            Transaction txn = transactions.get(i);
            String type = txn.getType();
            if (("TRANSFER".equals(type) || "CASH_OUT".equals(type)) && txn.getAmount() > 10000) {
                if ("TRANSFER".equals(type) && cashoutDests.contains(txn.getNameDest())) {
                    riskMap.put(i, new RiskResult(10.0,
                            "Large transfer to account that also performs cash-out (potential layering)"));
                } else if ("CASH_OUT".equals(type) && txn.getAmount() > 50000) {
                    riskMap.put(i, new RiskResult(10.0,
                            String.format("Large cash-out transaction (%,.2f), potential fraud cash-out",
                                    txn.getAmount())));
                }
            }
        }

        return riskMap;
    }

    /**
     * Assign risk category based on score thresholds.
     */
    public String assignRiskLevel(double score) {
        if (score > 70) {
            return "HIGH";
        } else if (score >= 40) {
            return "MEDIUM";
        }
        return "LOW";
    }

    /**
     * Generate a comprehensive transaction-level risk report.
     *
     * Combines all risk signals to compute a final risk score and category
     * for each transaction.
     */
    public List<RiskReport> generateRiskReport(List<Transaction> transactions) {
        // Pre-compute aggregate risk signals
        Map<Integer, RiskResult> repeatRisk = computeRepeatAccountRisk(transactions);
        Map<Integer, RiskResult> destRisk = computeDestinationRisk(transactions);
        Map<Integer, RiskResult> cashoutRisk = computeCashoutPatternRisk(transactions);

        List<RiskReport> results = new ArrayList<>();

        for (int i = 0; i < transactions.size(); i++) {
            Transaction txn = transactions.get(i);
            List<String> explanations = new ArrayList<>();
            double totalScore = 0.0;

            // 1. Amount risk
            RiskResult amtResult = computeAmountRisk(txn.getAmount());
            totalScore += amtResult.getScore();
            if (!amtResult.getReason().isEmpty()) {
                explanations.add(amtResult.getReason());
            }

            // 2. Transaction type risk
            RiskResult typeResult = computeTypeRisk(txn.getType());
            totalScore += typeResult.getScore();
            if (!typeResult.getReason().isEmpty()) {
                explanations.add(typeResult.getReason());
            }

            // 3. Balance anomaly risk
            RiskResult balResult = computeBalanceAnomalyRisk(txn);
            totalScore += balResult.getScore();
            if (!balResult.getReason().isEmpty()) {
                explanations.add(balResult.getReason());
            }

            // 4. Repeated account risk
            RiskResult repResult = repeatRisk.get(i);
            totalScore += repResult.getScore();
            if (!repResult.getReason().isEmpty()) {
                explanations.add(repResult.getReason());
            }

            // 5. Destination risk
            RiskResult dstResult = destRisk.get(i);
            totalScore += dstResult.getScore();
            if (!dstResult.getReason().isEmpty()) {
                explanations.add(dstResult.getReason());
            }

            // 6. Cash-out pattern risk
            RiskResult coResult = cashoutRisk.get(i);
            totalScore += coResult.getScore();
            if (!coResult.getReason().isEmpty()) {
                explanations.add(coResult.getReason());
            }

            // 7. Known fraud indicator boost
            if (txn.getIsFraud() == 1) {
                totalScore += 15.0;
                explanations.add("Transaction flagged as fraud in dataset");
            }
            if (txn.getIsFlaggedFraud() == 1) {
                totalScore += 10.0;
                explanations.add("Transaction flagged by business fraud detection rules");
            }

            // Cap score at 100
            double finalScore = Math.min(Math.round(totalScore * 100.0) / 100.0, 100.0);
            String riskLevel = assignRiskLevel(finalScore);
            String explanation = explanations.isEmpty()
                    ? "No significant risk signals detected"
                    : String.join("; ", explanations);

            results.add(new RiskReport(txn.getTransactionId(), finalScore, riskLevel, explanation));
        }

        return results;
    }

    /**
     * Save the risk report to a CSV file.
     */
    public void saveReport(List<RiskReport> reports, String outputPath) throws IOException {
        try (CSVWriter writer = new CSVWriter(new FileWriter(outputPath))) {
            writer.writeNext(new String[]{"transaction_id", "risk_score", "risk_level", "explanation"});
            for (RiskReport report : reports) {
                writer.writeNext(new String[]{
                        String.valueOf(report.getTransactionId()),
                        String.valueOf(report.getRiskScore()),
                        report.getRiskLevel(),
                        report.getExplanation()
                });
            }
        }
    }

    /**
     * Main entry point for the fraud risk scoring system.
     */
    public static void main(String[] args) {
        String inputPath = "data/Example1.csv";
        String outputPath = "data/fraud_risk_report.csv";

        FraudRiskScoring scorer = new FraudRiskScoring();

        try {
            System.out.printf("Loading dataset from '%s'...%n", inputPath);
            List<Transaction> transactions = scorer.loadDataset(inputPath);
            System.out.printf("Loaded %d transactions.%n", transactions.size());

            System.out.println("\nAnalyzing transactions for fraud risk signals...");
            List<RiskReport> report = scorer.generateRiskReport(transactions);

            // Summary statistics
            System.out.println("\n--- Risk Score Summary ---");
            System.out.printf("Total transactions analyzed: %d%n", report.size());

            // Risk level distribution
            Map<String, Integer> levelCounts = new HashMap<>();
            for (RiskReport r : report) {
                levelCounts.merge(r.getRiskLevel(), 1, Integer::sum);
            }
            System.out.println("Risk level distribution:");
            for (Map.Entry<String, Integer> entry : levelCounts.entrySet()) {
                System.out.printf("  %s: %d%n", entry.getKey(), entry.getValue());
            }

            // Risk score statistics
            double sum = 0, min = Double.MAX_VALUE, max = Double.MIN_VALUE;
            for (RiskReport r : report) {
                sum += r.getRiskScore();
                min = Math.min(min, r.getRiskScore());
                max = Math.max(max, r.getRiskScore());
            }
            double mean = report.isEmpty() ? 0 : sum / report.size();
            System.out.printf("%nRisk score statistics:%n");
            System.out.printf("  Mean: %.2f%n", mean);
            System.out.printf("  Min:  %.2f%n", min);
            System.out.printf("  Max:  %.2f%n", max);

            // Show HIGH risk transactions
            List<RiskReport> highRisk = new ArrayList<>();
            for (RiskReport r : report) {
                if ("HIGH".equals(r.getRiskLevel())) {
                    highRisk.add(r);
                }
            }
            if (!highRisk.isEmpty()) {
                System.out.printf("%n--- HIGH Risk Transactions (%d) ---%n", highRisk.size());
                for (RiskReport r : highRisk) {
                    System.out.printf("  Transaction %d: score=%.2f, explanation=%s%n",
                            r.getTransactionId(), r.getRiskScore(), r.getExplanation());
                }
            }

            // Save report
            scorer.saveReport(report, outputPath);
            System.out.printf("%nRisk report saved to '%s'%n", outputPath);

        } catch (Exception e) {
            System.err.println("Error: " + e.getMessage());
            e.printStackTrace();
        }
    }
}
